app/src/behavior_key.c

```diff
@@ -0,0 +1,159 @@
+    struct zmk_debounce_state debounce_state;
+    struct gpio_callback key_callback;
+    const struct device *dev;
+    struct k_work_delayable update_work;
+    uint32_t read_time;
+};
+
+static void bk_enable_interrupt(const struct device *dev) {
+    const struct behavior_key_config *config = dev->config;
+
+    gpio_pin_interrupt_configure_dt(&config->key, GPIO_INT_LEVEL_ACTIVE);
+}
+
+static void bk_disable_interrupt(const struct device *dev) {
+    const struct behavior_key_config *config = dev->config;
+
+    gpio_pin_interrupt_configure_dt(&config->key, GPIO_INT_DISABLE);
+}
+
+static void bk_read(const struct device *dev) {
+    const struct behavior_key_config *config = dev->config;
+    struct behavior_key_data *data = dev->data;
+
+    zmk_debounce_update(&data->debounce_state, gpio_pin_get_dt(&config->key),
+                        config->debounce_scan_period_ms, &config->debounce_config);
+
+    if (zmk_debounce_get_changed(&data->debounce_state)) {
+        const bool pressed = zmk_debounce_is_pressed(&data->debounce_state);
+
+        struct zmk_behavior_binding_event event = {.position = INT32_MAX,
+                                                   .timestamp = k_uptime_get()};
+
+        if (pressed) {
+            behavior_keymap_binding_pressed(&data->binding, event);
+        } else {
+            behavior_keymap_binding_released(&data->binding, event);
+        }
+    }
+
+    if (zmk_debounce_is_active(&data->debounce_state)) {
+        data->read_time += config->debounce_scan_period_ms;
+
+        k_work_reschedule(&data->update_work, K_TIMEOUT_ABS_MS(data->read_time));
+    } else {
+        bk_enable_interrupt(dev);
+    }
+}
+
+static void bk_update_work(struct k_work *work) {
+    struct k_work_delayable *dwork = CONTAINER_OF(work, struct k_work_delayable, work);
+    struct behavior_key_data *data = CONTAINER_OF(dwork, struct behavior_key_data, update_work);
+    bk_read(data->dev);
+}
+
+static void bk_gpio_irq_callback(const struct device *port, struct gpio_callback *cb,
+                                 const gpio_port_pins_t pin) {
+    struct behavior_key_data *data = CONTAINER_OF(cb, struct behavior_key_data, key_callback);
+
+    bk_disable_interrupt(data->dev);
+
+    data->read_time = k_uptime_get();
+    k_work_reschedule(&data->update_work, K_NO_WAIT);
+}
+
+static int behavior_key_init(const struct device *dev) {
+    const struct behavior_key_config *config = dev->config;
+    struct behavior_key_data *data = dev->data;
+
+    if (!device_is_ready(config->key.port)) {
+        LOG_ERR("GPIO port is not ready");
+        return -ENODEV;
+    }
+
+    k_work_init_delayable(&data->update_work, bk_update_work);
+    data->dev = dev;
+
+    gpio_pin_configure_dt(&config->key, GPIO_INPUT);
+    gpio_init_callback(&data->key_callback, bk_gpio_irq_callback, BIT(config->key.pin));
+    gpio_add_callback(config->key.port, &data->key_callback);
+
+    while (gpio_pin_get_dt(&config->key)) {
```

Comment:
This probably deserves a comment to explain why this is being done. I'm guessing this is to avoid the case where the keyboard just woke up due to pressing this key, so you don't want to immediately read that as a trigger to go back to sleep? (Does this need debouncing too? If so, moving this out to a "wait for release" function would be helpful to keep the init function simple.)

Reply:
Pulled this out into a separate function and added a comment. Thanks.